test.py

```diff
@@ -0,0 +1,311 @@
+
+    def test_from_good_to_dangerous(self):
+        a = Good()
+
+        with self.assertRaisesRegex(ValueError, "safe_eval: doesn't permit you to modify an attribute"):
+            exec(expr_checker("a.__class__ = Dangerous", ast_get_attr, check_type=check_type))
+
+    def test_hijacking_global(self):
+        with self.assertRaisesRegex(ValueError, "safe_eval: doesn't permit you to modify locals\(\) and globals\(\)"):
+            exec(expr_checker("globals()['Good'] = Dangerous", ast_get_attr, check_type=check_type))
+
+        with self.assertRaisesRegex(ValueError, "safe_eval: doesn't permit you to modify locals\(\) and globals\(\)"):
+            exec(expr_checker(cleandoc("locals()['evil'] = lambda: Dangerous()"), ast_get_attr, check_type=check_type))
+
+    def test_forbidden_attr(self):
+        a = Good()
+
+        with self.assertRaisesRegex(ValueError, "safe_eval doesn't permit you to read _secret_stuff"):
+            exec(expr_checker("print(a._secret_stuff)", ast_get_attr, check_type=check_type))
+
+        # FIXME with ast_set_attr
+        with self.assertRaisesRegex(ValueError, "safe_eval: doesn't permit you to modify an attribute"):
+            exec(expr_checker("a._secret_stuff = 42", ast_get_attr, check_type=check_type))
+
+    def test_dangerous_lambda(self):
+        with self.assertRaisesRegex(ValueError, "safe_eval didn't like <class '__main__.Dangerous'>"):
+            exec(expr_checker("(lambda: print(Dangerous()))()", ast_get_attr, check_type=check_type))
+
+    def test_multiline(self):
+        code = cleandoc(
+            """
+            def b():
+                print('.', end='')
+
+            b()
+            """
+        )
+
+        exec(expr_checker(code, ast_get_attr, check_type=check_type))
+
+    def test_file_open(self):
+        a = Good()
+        a.evil = open("test.py")
+
+        # FIXME with ast_set_attr
+        with self.assertRaisesRegex(ValueError, "safe_eval doesn't permit you to read evil"):
+            exec(expr_checker("print(a.evil)", ast_get_attr, check_type=check_type))
+
+        a.evil.close()
+
+    def test_overwrite(self):
+        a = Good()
+
+        with self.assertRaisesRegex(NameError, "safe_eval: check_type is a reserved name"):
+            code = cleandoc(
+                """
+                def check_type(t):
+                    return t
+
+                c = a.evil
+                """
+            )
+
+            exec(expr_checker(code, ast_get_attr, check_type=check_type))
+
+        with self.assertRaisesRegex(NameError, "safe_eval: check_type is a reserved name"):
+            code = cleandoc(
+                """
+                check_type = lambda t: t
+                c = a.evil
+                """
+            )
+
+            exec(expr_checker(code, ast_get_attr, check_type=check_type))
+
+    def test_isinstance_bad_idea(self):
+        class Dangerous2(Good):
```

Comment:
so you do not need to create `check_type2`
```suggestion
        class Dangerous2(int):
```

Reply:
Well, I need it for the little demo 😅 
`check_type2` uses `is_instance()` 
and  `check_type` uses `type()`

Reply:
yes I know, but it's not needed to test a wrong version of check_types :)